File: _engine/java/v0_1_0/Json.java
// Copyright 2026 GildenEye
// SPDX-License-Identifier: Apache-2.0

package aifunc._engine.java.v0_1_0;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Zero-dependency JSON parser and serializer.
 *
 * <p>Parsed values are represented as standard Java types:
 * <ul>
 *   <li>JSON object  → {@code Map<String, Object>}
 *   <li>JSON array   → {@code List<Object>}
 *   <li>JSON string  → {@code String}
 *   <li>JSON number  → {@code Double}
 *   <li>JSON boolean → {@code Boolean}
 *   <li>JSON null    → {@code null}
 * </ul>
 */
public final class Json {

    private Json() {}

    // -------------------------------------------------------------------------
    // Public API
    // -------------------------------------------------------------------------

    /** Parses a JSON string and returns the corresponding Java object. */
    public static Object parse(String json) {
        if (json == null) throw new JsonParseException("Input is null");
        Parser p = new Parser(json.trim());
        Object result = p.parseValue();
        p.skipWhitespace();
        if (p.pos < p.src.length()) {
            throw new JsonParseException("Unexpected trailing content at position " + p.pos);
        }
        return result;
    }

    /** Parses a JSON string as a {@code Map<String, Object>}. */
    @SuppressWarnings("unchecked")
    public static Map<String, Object> parseObject(String json) {
        Object v = parse(json);
        if (!(v instanceof Map)) {
            throw new JsonParseException("Expected JSON object, got: " + typeName(v));
        }
        return (Map<String, Object>) v;
    }

    /** Serializes a Java value to a compact JSON string. */
    public static String stringify(Object value) {
        StringBuilder sb = new StringBuilder();
        writeValue(value, sb, -1);
        return sb.toString();
    }

    /** Serializes a Java value to a pretty-printed JSON string with 2-space indentation. */
    public static String prettyPrint(Object value) {
        StringBuilder sb = new StringBuilder();
        writeValue(value, sb, 0);
        return sb.toString();
    }

    // -------------------------------------------------------------------------
    // Serializer
    // -------------------------------------------------------------------------

    @SuppressWarnings("unchecked")
    private static void writeValue(Object value, StringBuilder sb, int indent) {
        if (value == null) {
            sb.append("null");
        } else if (value instanceof Boolean) {
            sb.append((Boolean) value ? "true" : "false");
        } else if (value instanceof Number) {
            double d = ((Number) value).doubleValue();
            if (d == Math.floor(d) && !Double.isInfinite(d) && Math.abs(d) < 1e15) {
                sb.append((long) d);
            } else {
                sb.append(d);
            }
        } else if (value instanceof String) {
            writeString((String) value, sb);
        } else if (value instanceof List) {
            writeList((List<Object>) value, sb, indent);
        } else if (value instanceof Map) {
            writeMap((Map<String, Object>) value, sb, indent);
        } else {
            writeString(value.toString(), sb);
        }
    }

    private static void writeString(String s, StringBuilder sb) {
        sb.append('"');
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            switch (c) {
                case '"':  sb.append("\\\""); break;
                case '\\': sb.append("\\\\"); break;
                case '\b': sb.append("\\b");  break;
                case '\f': sb.append("\\f");  break;
                case '\n': sb.append("\\n");  break;
                case '\r': sb.append("\\r");  break;
                case '\t': sb.append("\\t");  break;
                default:
                    if (c < 0x20) {
                        sb.append(String.format("\\u%04x", (int) c));
                    } else {
                        sb.append(c);
                    }
            }
        }
        sb.append('"');
    }

    private static void writeList(List<Object> list, StringBuilder sb, int indent) {
        if (list.isEmpty()) { sb.append("[]"); return; }
        sb.append('[');
        int child = indent >= 0 ? indent + 1 : -1;
        for (int i = 0; i < list.size(); i++) {
            if (indent >= 0) { sb.append('\n'); appendIndent(sb, child); }
            writeValue(list.get(i), sb, child);
            if (i < list.size() - 1) sb.append(',');
        }
        if (indent >= 0) { sb.append('\n'); appendIndent(sb, indent); }
        sb.append(']');
    }

    private static void writeMap(Map<String, Object> map, StringBuilder sb, int indent) {
        if (map.isEmpty()) { sb.append("{}"); return; }
        sb.append('{');
        int child = indent >= 0 ? indent + 1 : -1;
        int i = 0;
        for (Map.Entry<String, Object> entry : map.entrySet()) {
            if (indent >= 0) { sb.append('\n'); appendIndent(sb, child); }
            writeString(entry.getKey(), sb);
            sb.append(':');
            if (indent >= 0) sb.append(' ');
            writeValue(entry.getValue(), sb, child);
            if (i < map.size() - 1) sb.append(',');
            i++;
        }
        if (indent >= 0) { sb.append('\n'); appendIndent(sb, indent); }
        sb.append('}');
    }

    private static void appendIndent(StringBuilder sb, int level) {
        for (int i = 0; i < level * 2; i++) sb.append(' ');
    }

    // -------------------------------------------------------------------------
    // Parser (recursive descent)
    // -------------------------------------------------------------------------

    static final class Parser {
        final String src;
        int pos;

        Parser(String src) { this.src = src; this.pos = 0; }

        Object parseValue() {
            skipWhitespace();
            if (pos >= src.length()) throw new JsonParseException("Unexpected end of input");
            char c = src.charAt(pos);
            if (c == '"') return parseString();
            if (c == '{') return parseObject();
            if (c == '[') return parseArray();
            if (c == 't') return parseLiteral("true",  Boolean.TRUE);
            if (c == 'f') return parseLiteral("false", Boolean.FALSE);
            if (c == 'n') return parseLiteral("null",  null);
            if (c == '-' || (c >= '0' && c <= '9')) return parseNumber();
            throw new JsonParseException("Unexpected character '" + c + "' at position " + pos);
        }

        private Map<String, Object> parseObject() {
            expect('{');
            Map<String, Object> map = new LinkedHashMap<>();
            skipWhitespace();
            if (peek() == '}') { pos++; return map; }
            while (true) {
                skipWhitespace();
                String key = parseString();
                skipWhitespace(); expect(':');
                Object value = parseValue();
                map.put(key, value);
                skipWhitespace();
                char sep = peek();
                if (sep == '}') { pos++; break; }
                if (sep == ',') { pos++; continue; }
                throw new JsonParseException("Expected ',' or '}' at position " + pos);
            }
            return map;
        }

        private List<Object> parseArray() {
            expect('[');
            List<Object> list = new ArrayList<>();
            skipWhitespace();
            if (peek() == ']') { pos++; return list; }
            while (true) {
                list.add(parseValue());
                skipWhitespace();
                char sep = peek();
                if (sep == ']') { pos++; break; }
                if (sep == ',') { pos++; continue; }
                throw new JsonParseException("Expected ',' or ']' at position " + pos);
            }
            return list;
        }

        String parseString() {
            skipWhitespace(); expect('"');
            StringBuilder sb = new StringBuilder();
            while (pos < src.length()) {
                char c = src.charAt(pos++);
                if (c == '"') return sb.toString();
                if (c != '\\') { sb.append(c); continue; }
                if (pos >= src.length()) break;
                char esc = src.charAt(pos++);
                switch (esc) {
                    case '"':  sb.append('"');  break;
                    case '\\': sb.append('\\'); break;
                    case '/':  sb.append('/');  break;
                    case 'b':  sb.append('\b'); break;
                    case 'f':  sb.append('\f'); break;
                    case 'n':  sb.append('\n'); break;
                    case 'r':  sb.append('\r'); break;
                    case 't':  sb.append('\t'); break;
                    case 'u':
                        if (pos + 4 > src.length())
                            throw new JsonParseException("Incomplete \\u escape");
                        sb.append((char) Integer.parseInt(src.substring(pos, pos + 4), 16));
                        pos += 4; break;
                    default:
                        throw new JsonParseException("Invalid escape '\\" + esc + "'");
                }
            }
            throw new JsonParseException("Unterminated string");
        }

        private Double parseNumber() {
            int start = pos;
            if (pos < src.length() && src.charAt(pos) == '-') pos++;
            while (pos < src.length() && src.charAt(pos) >= '0' && src.charAt(pos) <= '9') pos++;
            if (pos < src.length() && src.charAt(pos) == '.') {
                pos++;
                while (pos < src.length() && src.charAt(pos) >= '0' && src.charAt(pos) <= '9') pos++;
            }
            if (pos < src.length() && (src.charAt(pos) == 'e' || src.charAt(pos) == 'E')) {
                pos++;
                if (pos < src.length() && (src.charAt(pos) == '+' || src.charAt(pos) == '-')) pos++;
                while (pos < src.length() && src.charAt(pos) >= '0' && src.charAt(pos) <= '9') pos++;
            }
            try {
                return Double.parseDouble(src.substring(start, pos));
            } catch (NumberFormatException e) {
                throw new JsonParseException("Invalid number at position " + start);
            }
        }

        private Object parseLiteral(String literal, Object value) {
            if (src.startsWith(literal, pos)) { pos += literal.length(); return value; }
            throw new JsonParseException("Expected '" + literal + "' at position " + pos);
        }

        void skipWhitespace() {
            while (pos < src.length()) {
                char c = src.charAt(pos);
                if (c == ' ' || c == '\t' || c == '\n' || c == '\r') pos++;
                else break;
            }
        }

        private char peek() {
            if (pos >= src.length())
                throw new JsonParseException("Unexpected end of input");
            return src.charAt(pos);
        }

        private void expect(char c) {
            skipWhitespace();
            if (pos >= src.length() || src.charAt(pos) != c) {
                String got = pos < src.length() ? String.valueOf(src.charAt(pos)) : "EOF";
                throw new JsonParseException("Expected '" + c + "' but got '" + got + "' at position " + pos);
            }
            pos++;
        }
    }

    private static String typeName(Object v) {
        if (v == null) return "null";
        if (v instanceof Map) return "object";
        if (v instanceof List) return "array";
        if (v instanceof String) return "string";
        if (v instanceof Number) return "number";
        if (v instanceof Boolean) return "boolean";
        return v.getClass().getSimpleName();
    }

    // -------------------------------------------------------------------------
    // Exception
    // -------------------------------------------------------------------------

    /** Thrown when the JSON input cannot be parsed. */
    public static final class JsonParseException extends RuntimeException {
        public JsonParseException(String message) { super(message); }
    }
}

File: _engine/csharp/v0_2_0/Json.cs
// Copyright 2026 GildenEye
// SPDX-License-Identifier: Apache-2.0

using System;
using System.Collections;
using System.Collections.Generic;
using System.Globalization;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace Aifunc.Engine.Csharp.V0_2_0;

/// <summary>
/// Zero-dependency JSON parser and serializer built on System.Text.Json.Nodes.
///
/// Parsed values are represented as standard BCL types:
/// <list type="bullet">
///   <item>JSON object  â†?<c>Dictionary&lt;string, object?&gt;</c></item>
///   <item>JSON array   â†?<c>List&lt;object?&gt;</c></item>
///   <item>JSON string  â†?<c>string</c></item>
///   <item>JSON number  â†?<c>double</c></item>
///   <item>JSON boolean â†?<c>bool</c></item>
///   <item>JSON null    â†?<c>null</c></item>
/// </list>
/// </summary>
public static class Json
{
    /// <summary>Parses a JSON string and returns the corresponding BCL object.</summary>
    public static object? Parse(string json)
    {
        if (json is null) throw new JsonParseException("Input is null");
        try
        {
            var node = JsonNode.Parse(json.Trim());
            return FromNode(node);
        }
        catch (JsonException e)
        {
            throw new JsonParseException(e.Message);
        }
    }

    /// <summary>Parses a JSON string as a <c>Dictionary&lt;string, object?&gt;</c>.</summary>
    public static Dictionary<string, object?> ParseObject(string json)
    {
        var v = Parse(json);
        if (v is not Dictionary<string, object?> map)
        {
            throw new JsonParseException("Expected JSON object, got: " + TypeName(v));
        }
        return map;
    }

    /// <summary>Serializes a BCL value to a compact JSON string.</summary>
    public static string Stringify(object? value)
    {
        var sb = new StringBuilder();
        WriteValue(value, sb, -1);
        return sb.ToString();
    }

    /// <summary>Serializes a BCL value to a pretty-printed JSON string with 2-space indentation.</summary>
    public static string PrettyPrint(object? value)
    {
        var sb = new StringBuilder();
        WriteValue(value, sb, 0);
        return sb.ToString();
    }

    // -------------------------------------------------------------------------
    // Node â†?BCL
    // -------------------------------------------------------------------------

    private static object? FromNode(JsonNode? node)
    {
        if (node is null) return null;

        if (node is JsonObject obj)
        {
            var map = new Dictionary<string, object?>();
            foreach (var prop in obj)
            {
                map[prop.Key] = FromNode(prop.Value);
            }
            return map;
        }

        if (node is JsonArray arr)
        {
            var list = new List<object?>();
            foreach (var item in arr)
            {
                list.Add(FromNode(item));
            }
            return list;
        }

        if (node is JsonValue val)
        {
            // Prefer explicit typed getters so numbers always become double.
            if (val.TryGetValue<bool>(out var b)) return b;
            if (val.TryGetValue<string>(out var s)) return s;
            if (val.TryGetValue<double>(out var d)) return d;
            // JsonValue may hold JsonElement for some payloads.
            var element = val.GetValue<object?>();
            if (element is JsonElement je) return FromElement(je);
            if (element is null) return null;
            if (element is bool bb) return bb;
            if (element is string ss) return ss;
            if (element is IConvertible)
            {
                try { return Convert.ToDouble(element, CultureInfo.InvariantCulture); }
                catch { /* fall through */ }
            }
            return element.ToString();
        }

        return null;
    }

    private static object? FromElement(JsonElement el)
    {
        switch (el.ValueKind)
        {
            case JsonValueKind.Null:
            case JsonValueKind.Undefined:
                return null;
            case JsonValueKind.True:
                return true;
            case JsonValueKind.False:
                return false;
            case JsonValueKind.String:
                return el.GetString();
            case JsonValueKind.Number:
                return el.GetDouble();
            case JsonValueKind.Object:
            {
                var map = new Dictionary<string, object?>();
                foreach (var prop in el.EnumerateObject())
                {
                    map[prop.Name] = FromElement(prop.Value);
                }
                return map;
            }
            case JsonValueKind.Array:
            {
                var list = new List<object?>();
                foreach (var item in el.EnumerateArray())
                {
                    list.Add(FromElement(item));
                }
                return list;
            }
            default:
                return null;
        }
    }

    // -------------------------------------------------------------------------
    // Serializer
    // -------------------------------------------------------------------------

    private static void WriteValue(object? value, StringBuilder sb, int indent)
    {
        if (value is null)
        {
            sb.Append("null");
        }
        else if (value is bool b)
        {
            sb.Append(b ? "true" : "false");
        }
        else if (value is byte or sbyte or short or ushort or int or uint or long or ulong
                 or float or double or decimal)
        {
            var d = Convert.ToDouble(value, CultureInfo.InvariantCulture);
            if (d == Math.Floor(d) && !double.IsInfinity(d) && Math.Abs(d) < 1e15)
            {
                sb.Append((long)d);
            }
            else
            {
                sb.Append(d.ToString(CultureInfo.InvariantCulture));
            }
        }
        else if (value is string s)
        {
            WriteString(s, sb);
        }
        else if (value is IList<object?> list)
        {
            WriteList(list, sb, indent);
        }
        else if (value is System.Collections.IList rawList)
        {
            var converted = new List<object?>();
            foreach (var item in rawList) converted.Add(item);
            WriteList(converted, sb, indent);
        }
        else if (value is IDictionary<string, object?> map)
        {
            WriteMap(map, sb, indent);
        }
        else if (value is System.Collections.IDictionary rawMap)
        {
            var converted = new Dictionary<string, object?>();
            foreach (System.Collections.DictionaryEntry entry in rawMap)
            {
                converted[Convert.ToString(entry.Key, CultureInfo.InvariantCulture) ?? ""] = entry.Value;
            }
            WriteMap(converted, sb, indent);
        }
        else
        {
            WriteString(value.ToString() ?? "", sb);
        }
    }

    private static void WriteString(string s, StringBuilder sb)
    {
        sb.Append('"');
        foreach (var c in s)
        {
            switch (c)
            {
                case '"':  sb.Append("\\\""); break;
                case '\\': sb.Append("\\\\"); break;
                case '\b': sb.Append("\\b");  break;
                case '\f': sb.Append("\\f");  break;
                case '\n': sb.Append("\\n");  break;
                case '\r': sb.Append("\\r");  break;
                case '\t': sb.Append("\\t");  break;
                default:
                    if (c < 0x20)
                    {
                        sb.Append("\\u");
                        sb.Append(((int)c).ToString("x4", CultureInfo.InvariantCulture));
                    }
                    else
                    {
                        sb.Append(c);
                    }
                    break;
            }
        }
        sb.Append('"');
    }

    private static void WriteList(IList<object?> list, StringBuilder sb, int indent)
    {
        if (list.Count == 0) { sb.Append("[]"); return; }
        sb.Append('[');
        var child = indent >= 0 ? indent + 1 : -1;
        for (var i = 0; i < list.Count; i++)
        {
            if (indent >= 0) { sb.Append('\n'); AppendIndent(sb, child); }
            WriteValue(list[i], sb, child);
            if (i < list.Count - 1) sb.Append(',');
        }
        if (indent >= 0) { sb.Append('\n'); AppendIndent(sb, indent); }
        sb.Append(']');
    }

    private static void WriteMap(IDictionary<string, object?> map, StringBuilder sb, int indent)
    {
        if (map.Count == 0) { sb.Append("{}"); return; }
        sb.Append('{');
        var child = indent >= 0 ? indent + 1 : -1;
        var i = 0;
        foreach (var entry in map)
        {
            if (indent >= 0) { sb.Append('\n'); AppendIndent(sb, child); }
            WriteString(entry.Key, sb);
            sb.Append(':');
            if (indent >= 0) sb.Append(' ');
            WriteValue(entry.Value, sb, child);
            if (i < map.Count - 1) sb.Append(',');
            i++;
        }
        if (indent >= 0) { sb.Append('\n'); AppendIndent(sb, indent); }
        sb.Append('}');
    }

    private static void AppendIndent(StringBuilder sb, int level)
    {
        for (var i = 0; i < level * 2; i++) sb.Append(' ');
    }

    private static string TypeName(object? v)
    {
        if (v is null) return "null";
        if (v is IDictionary<string, object?> or System.Collections.IDictionary) return "object";
        if (v is IList<object?> or System.Collections.IList) return "array";
        if (v is string) return "string";
        if (v is byte or sbyte or short or ushort or int or uint or long or ulong
            or float or double or decimal) return "number";
        if (v is bool) return "boolean";
        return v.GetType().Name;
    }

    /// <summary>Thrown when the JSON input cannot be parsed.</summary>
    public sealed class JsonParseException : Exception
    {
        public JsonParseException(string message) : base(message) {}
    }
}
